article_maker.cc

```diff
@@ -642,19 +661,56 @@ void ArticleRequest::bodyFinished()
 
         closePrevSpan = true;
 
-        head += string( R"(<div class="gddictname" onclick="gdExpandArticle(')" ) + dictId + "\');"
-          + ( collapse ? "\" style=\"cursor:pointer;" : "" )
-          + "\" id=\"gddictname-" + Html::escape( dictId ) + "\""
-          + ( collapse ? string( " title=\"" ) + tr( "Expand article" ).toUtf8().data() + "\"" : "" )
-          + R"(><span class="gddicticon"><img src="gico://)" + Html::escape( dictId )
-          + R"(/dicticon.png"></span><span class="gdfromprefix">)"  +
-          Html::escape( tr( "From " ).toUtf8().data() ) + "</span><span class=\"gddicttitle\">" +
-          Html::escape( activeDict->getName().c_str() ) + "</span>"
-          + R"(<span class="collapse_expand_area"><img src="qrc:///icons/blank.png" class=")"
-          + ( collapse ? "gdexpandicon" : "gdcollapseicon" )
-          + "\" id=\"expandicon-" + Html::escape( dictId ) + "\""
-          + ( collapse ? "" : string( " title=\"" ) + tr( "Collapse article" ).toUtf8().data() + "\"" )
-          + "></span>" + "</div>";
+        // Start .gddictname
+        // Revert to QString here to be able to use its arg() method.
+        auto const dict_id_html = Html::escape( dictId );
+        QString gd_dict_name    = R"EOF(<div class="gddictname" id="gddictname-%1">)EOF";
+        gd_dict_name            = gd_dict_name.arg( dict_id_html.c_str() );
+
+        // Icon
+        gd_dict_name += R"EOF(
+        <span class="gddicticon"><img src="gico://%1/dicticon.png"></span>
+        <span class="gdfromprefix">%2</span>
+        <span class="gddicttitle">%3</span>
+        )EOF";
+        gd_dict_name = gd_dict_name.arg(
+          dict_id_html.c_str(),
+          tr( "From " ),
+          activeDict->getName().c_str()
+        );
+
+        // Collapse/expand button (blue arrow)
+        gd_dict_name += R"EOF(
+        <span class="collapse_expand_area" onclick="gdExpandArticle('%1');" >
+        <img src="qrc:///icons/blank.png" class="%2" id="expandicon-%3" title="%4">
+        </span>
+        )EOF";
+        gd_dict_name = gd_dict_name.arg(
+          dict_id_html.c_str(),
+          collapse ? "gdexpandicon" : "gdcollapseicon",
+          dict_id_html.c_str(),
+          collapse ? tr( "Expand article" ) : tr( "Collapse article" )
+        );
+
+        // If the user has enabled Anki integration in settings,
+        // Show a (+) button that lets the user add a new Anki card.
+        if ( ankiConnectEnabled ) {
+          gd_dict_name += R"EOF(
+          <a href="ankicard:%1" class="ankibutton" title="%2" >
+          <img src="qrc:///icons/add-anki-icon.svg">
+          </a>
+          )EOF";
+          gd_dict_name = gd_dict_name.arg(
+            dict_id_html.c_str(),
+            tr( "Make a new Anki note" )
+          );
+        }
```

Comment:
this can use a style to hold the place and use css to control the display. pseudocode:
html:
<div class=anki-button> </div>
css:
.anki-button{
background-img:xxx
}

The following is my guess(choose if it is easier and suitable)
there is no neccessary to pass dict_id to the ankicard:%s.
check the function about how to select current dictionary .   When click the button ,get the dictionary id which hold this button .which will need a javascript event listener for .ankibutton

Reply:
> this can use a style to hold the place and use css to control the display.

The above code uses `<img>` inside a `<span>` which seems reasonable.

```
<span class="collapse_expand_area" onclick="gdExpandArticle('%1');" >
<img src="qrc:///icons/blank.png" class="%2" id="expandicon-%3" title="%4">
</span>
```

For the anki add button we decided to follow the same pattern. Of course, alternatively it's possible to have `"qrc:///icons/add-anki-icon.svg` as a background image, but then we'd have to explicitly style the `<a>` tag, adding quite a lot of additional CSS (size, background position, repeat, etc.). I have no preference but it seems the current solution is easier.

However, I think it's a good idea to use css to control the visibilty (`display:none` if ankiconnect is disabled). I'll try to implement that.

> there is no neccessary to pass dict_id to the ankicard:%s.

I think you're right, but `dict_id_html` is already used quite a number of times throughout the code so it's easier to pass it right away instead of fetching the article's id later with additional JavaScript.

Reply:
> However, I think it's a good idea to use css to control the visibilty (`display:none` if ankiconnect is disabled). I'll try to implement that.

I think if not enabled,do not output the placeholder